#### Description

There are 2 datasets present in the file. Data 1 and Data 2
The primary key for both data1 and data2 is Order Id + Product ID combination (i.e. the individual datasets do not have any duplicate on this combination)


In [1]:
## loading the data

In [2]:
import pandas as pd

In [5]:
data1 = pd.read_excel(r"d:\Users\HP\Desktop\Siddhant Jain\NeenOpal\Records Matching Task.xlsx",
                      sheet_name = 'data1',
                      header = 0)
data1.head()

Order ID       Product ID  Qty
0  CA-2014-100006  TEC-PH-10002075    3
1  CA-2014-100090  FUR-TA-10003715    3
2  CA-2014-100090  OFF-BI-10001597    6
3  CA-2014-100293  OFF-PA-10000176    6
4  CA-2014-100328  OFF-BI-10000343    1

In [10]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9478 entries, 0 to 9477
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Order ID    9478 non-null   object
 1   Product ID  9478 non-null   object
 2   Qty         9478 non-null   int64 
dtypes: int64(1), object(2)
memory usage: 222.3+ KB


In [14]:
data1.isnull().sum()

Order ID      0
Product ID    0
Qty           0
dtype: int64

In [6]:
data2 = pd.read_excel(r"d:\Users\HP\Desktop\Siddhant Jain\NeenOpal\Records Matching Task.xlsx",
                      sheet_name = 'data2',
                      header = 0)
data2.head()

Order ID       Product ID  Qty
0  CA-2014-100090  FUR-TA-10003715    3
1  CA-2014-100090  OFF-BI-10001597    6
2  CA-2014-100293  OFF-PA-10000176    6
3  CA-2014-100328  OFF-BI-10000343    1
4  CA-2014-100363  OFF-FA-10000611    2

In [12]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9479 entries, 0 to 9478
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Order ID    9479 non-null   object
 1   Product ID  9479 non-null   object
 2   Qty         9479 non-null   int64 
dtypes: int64(1), object(2)
memory usage: 222.3+ KB


In [15]:
data2.isnull().sum()

Order ID      0
Product ID    0
Qty           0
dtype: int64

#### Provide solution and approach for the following:

#### 1. How to identify the Records (Order ID + Product ID combination) present in data1 but missing in data2 (Specify the number of records missing in your answer)

In [16]:
## merging the 2 data sets

In [75]:
data_merge_1 = pd.merge(left=data1,right=data2,on=['Order ID','Product ID'],how='left',indicator=True,suffixes=['- data1','- data2'])

In [77]:
data_merge_1.head()

Order ID       Product ID  Qty- data1  Qty- data2     _merge
0  CA-2014-100006  TEC-PH-10002075           3         NaN  left_only
1  CA-2014-100090  FUR-TA-10003715           3         3.0       both
2  CA-2014-100090  OFF-BI-10001597           6         6.0       both
3  CA-2014-100293  OFF-PA-10000176           6         6.0       both
4  CA-2014-100328  OFF-BI-10000343           1         1.0       both

In [78]:
data_merge_1[data_merge_1['_merge'] == 'left_only']

Order ID       Product ID  Qty- data1  Qty- data2     _merge
0     CA-2014-100006  TEC-PH-10002075           3         NaN  left_only
10    CA-2014-100678  OFF-EN-10000056           3         NaN  left_only
19    CA-2014-100895  OFF-AR-10004511           2         NaN  left_only
35    CA-2014-101560  OFF-BI-10000309           3         NaN  left_only
61    CA-2014-102673  OFF-LA-10001771          12         NaN  left_only
...              ...              ...         ...         ...        ...
9390  US-2017-160836  OFF-AP-10001626           2         NaN  left_only
9403  US-2017-162558  FUR-FU-10002364           2         NaN  left_only
9420  US-2017-163657  OFF-BI-10000138           5         NaN  left_only
9427  US-2017-164056  FUR-TA-10001307           5         NaN  left_only
9435  US-2017-165456  FUR-CH-10003981           6         NaN  left_only

[507 rows x 5 columns]

In [79]:
data_merge_1[data_merge_1['_merge'] == 'left_only'].count()

Order ID      507
Product ID    507
Qty- data1    507
Qty- data2      0
_merge        507
dtype: int64

##### the number of records missing = 507

#### 2. How to identify the Records (Order ID + Product ID combination) missing in data1 but present in data2 (Specify the number of records missing in your answer)

In [80]:
data_merge_2 = pd.merge(left=data1,right=data2,on=['Order ID','Product ID'],how='right',indicator=True,suffixes=['- data1','- data2'])

In [81]:
data_merge_2.head(50)

Order ID       Product ID  Qty- data1  Qty- data2      _merge
0   CA-2014-100090  FUR-TA-10003715         3.0           3        both
1   CA-2014-100090  OFF-BI-10001597         6.0           6        both
2   CA-2014-100293  OFF-PA-10000176         6.0           6        both
3   CA-2014-100328  OFF-BI-10000343         1.0           1        both
4   CA-2014-100363  OFF-FA-10000611         2.0           2        both
5   CA-2014-100363  OFF-PA-10004733         3.0           3        both
6   CA-2014-100391  OFF-PA-10001471         2.0           2        both
7   CA-2014-100678  FUR-CH-10002602         3.0           3        both
8   CA-2014-100678  OFF-AR-10001868         2.0           2        both
9   CA-2014-100678  TEC-AC-10000474         3.0           3        both
10  CA-2014-100706  FUR-FU-10002268         6.0           6        both
11  CA-2014-100706  TEC-AC-10001314         NaN           2  right_only
12  CA-2014-100762  OFF-AR-10000380         4.0           4        both
13  CA-2014-100762  OFF-LA-10003930         2.0           2        both
14  CA-2014-100762  OFF-PA-10001815         NaN           3  right_only
15  CA-2014-100762  OFF-PA-10004082         2.0           2        both
16  CA-2014-100860  OFF-LA-10001982         5.0           5        both
17  CA-2014-100867  TEC-PH-10004922         6.0           6        both
18  CA-2014-100881  TEC-PH-10003273         3.0           3        both
19  CA-2014-100895  OFF-ST-10001490         2.0           2        both
20  CA-2014-100895  TEC-PH-10001425         3.0           3        both
21  CA-2014-100916  FUR-TA-10004607         4.0           4        both
22  CA-2014-100916  OFF-AR-10001022         1.0           1        both
23  CA-2014-100916  OFF-ST-10003479         5.0           5        both
24  CA-2014-100972  OFF-PA-10000357         3.0           3        both
25  CA-2014-101147  OFF-AP-10004249         1.0           1        both
26  CA-2014-101175  OFF-ST-10004950         6.0           6        both
27  CA-2014-101266  OFF-PA-10002986         2.0           2        both
28  CA-2014-101364  OFF-BI-10003984        13.0          13        both
29  CA-2014-101392  OFF-ST-10000532         7.0           7        both
30  CA-2014-101427  OFF-AR-10002257         NaN           3  right_only
31  CA-2014-101462  FUR-FU-10000409         4.0           4        both
32  CA-2014-101476  TEC-MA-10000029         1.0           1        both
33  CA-2014-101560  FUR-FU-10003773         5.0           5        both
34  CA-2014-101560  OFF-AR-10004757         4.0           4        both
35  CA-2014-101560  OFF-PA-10001363         7.0           7        both
36  CA-2014-101602  FUR-CH-10004675         5.0           5        both
37  CA-2014-101602  TEC-PH-10000169         3.0           3        both
38  CA-2014-101770  OFF-BI-10001097         1.0           1        both
39  CA-2014-101833  OFF-PA-10003591         3.0           3        both
40  CA-2014-101931  FUR-BO-10001337         6.0           6        both
41  CA-2014-101931  OFF-BI-10004140         2.0           2        both
42  CA-2014-101931  OFF-ST-10003442         5.0           5        both
43  CA-2014-101931  OFF-SU-10000646         2.0           2        both
44  CA-2014-101931  OFF-SU-10002301         2.0           2        both
45  CA-2014-102008  OFF-PA-10004092         1.0           1        both
46  CA-2014-102085  OFF-LA-10002475         4.0           4        both
47  CA-2014-102274  OFF-AR-10002833         3.0           3        both
48  CA-2014-102274  OFF-PA-10004359         6.0           6        both
49  CA-2014-102274  OFF-ST-10001511         3.0           3        both

In [82]:
data_merge_2[data_merge_2['_merge'] == 'right_only']

Order ID       Product ID  Qty- data1  Qty- data2      _merge
11    CA-2014-100706  TEC-AC-10001314         NaN           2  right_only
14    CA-2014-100762  OFF-PA-10001815         NaN           3  right_only
30    CA-2014-101427  OFF-AR-10002257         NaN           3  right_only
56    CA-2014-102652  FUR-FU-10001918         NaN           7  right_only
63    CA-2014-102869  OFF-PA-10000788         NaN           3  right_only
...              ...              ...         ...         ...         ...
9428  US-2017-165344  OFF-BI-10003196         NaN          10  right_only
9433  US-2017-165358  TEC-CO-10001943         NaN           5  right_only
9455  US-2017-167920  OFF-AP-10000159         NaN           5  right_only
9471  US-2017-169502  OFF-AP-10001947         NaN           5  right_only
9473  US-2017-169551  FUR-BO-10001519         NaN           3  right_only

[508 rows x 5 columns]

In [84]:
data_merge_2[data_merge_2['_merge'] == 'right_only'].count()

Order ID      508
Product ID    508
Qty- data1      0
Qty- data2    508
_merge        508
dtype: int64

##### the number of records missing = 508

#### 3. Find the Sum of the total Qty of Records missing in data1 but present in data2

In [85]:
dq = data_merge_2[data_merge_2['_merge'] == 'right_only']

In [87]:
dq['Qty- data2'].sum()

1956

##### Sum of the total Qty of Records missing in data1 but present in data2 = 1956

#### 4. Find the total number of unique records (Order ID + Product ID combination) present in the combined dataset of data1 and data2

In [88]:
data_merge_unique = pd.merge(left=data1,right=data2,on=['Order ID','Product ID'],how='outer',indicator=True)

In [92]:
data_merge_unique['_merge'].value_counts()

both          8971
right_only     508
left_only      507
Name: _merge, dtype: int64

#####  total number of unique records = 8971